libbeat/outputs/shipper/shipper.go

```diff
@@ -70,52 +85,59 @@ func makeShipper(
 		return outputs.Fail(err)
 	}
 
-	s := outputs.WithBackoff(&shipper{
+	s := &shipper{
 		log:      logp.NewLogger("shipper"),
 		observer: observer,
 		config:   config,
-		timeout:  config.Timeout,
-	}, config.Backoff.Init, config.Backoff.Max)
+	}
+
+	// for `Close` function to stop all the background work like acknowledgment loop
+	s.backgroundCtx, s.backgroundCancel = context.WithCancel(context.Background())
+
+	swb := outputs.WithBackoff(s, config.Backoff.Init, config.Backoff.Max)
 
-	return outputs.Success(config.BulkMaxSize, config.MaxRetries, s)
+	return outputs.Success(config.BulkMaxSize, config.MaxRetries, swb)
 }
 
 // Connect establishes connection to the shipper server and implements `outputs.Connectable`.
-func (c *shipper) Connect() error {
-	tls, err := tlscommon.LoadTLSConfig(c.config.TLS)
+func (s *shipper) Connect() error {
+	tls, err := tlscommon.LoadTLSConfig(s.config.TLS)
 	if err != nil {
 		return fmt.Errorf("invalid shipper TLS configuration: %w", err)
 	}
 
 	var creds credentials.TransportCredentials
-	if c.config.TLS != nil && c.config.TLS.Enabled != nil && *c.config.TLS.Enabled {
+	if s.config.TLS != nil && s.config.TLS.Enabled != nil && *s.config.TLS.Enabled {
 		creds = credentials.NewTLS(tls.ToConfig())
 	} else {
 		creds = insecure.NewCredentials()
 	}
 
 	opts := []grpc.DialOption{
 		grpc.WithConnectParams(grpc.ConnectParams{
-			MinConnectTimeout: c.config.Timeout,
+			MinConnectTimeout: s.config.Timeout,
 		}),
 		grpc.WithBlock(),
 		grpc.WithTransportCredentials(creds),
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), c.config.Timeout)
+	ctx, cancel := context.WithTimeout(context.Background(), s.config.Timeout)
 	defer cancel()
 
-	c.log.Debugf("trying to connect to %s...", c.config.Server)
+	s.log.Debugf("trying to connect to %s...", s.config.Server)
 
-	conn, err := grpc.DialContext(ctx, c.config.Server, opts...)
+	conn, err := grpc.DialContext(ctx, s.config.Server, opts...)
 	if err != nil {
 		return fmt.Errorf("shipper connection failed with: %w", err)
 	}
 
-	c.conn = conn
-	c.client = sc.NewProducerClient(conn)
+	s.conn = conn
+	s.clientMutex.Lock()
+	defer s.clientMutex.Unlock()
 
-	indexClient, err := c.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
+	s.client = sc.NewProducerClient(conn)
+
+	indexClient, err := s.client.PersistedIndex(ctx, &messages.PersistedIndexRequest{
```

Comment:
How about we make this the persistent rpc client, and pass it into `ackLoop`?
- `ackLoop` can ignore the uuid mismatch check and startup errors since it's a persistent connection and we already got the initial response so the uuid can never change
- `pendingBatch` can remove the `serverID` field since all pending batches will always have the same serverID as `shipper` -- the only way to get a different id is in a publish response, which immediately cancels all outstanding batches

Reply:
I think (1) is a very nice idea. However, I disagree with (2): what if reconnect to the same server? Would not it be more efficient to acknowledge pending batches without resending them? It could be just a temporary network issue and the server kept running and persisting our previously sent batches. Sending them again would cause data duplication too.

I'd rather track the IDs and keep these batches in the pending list for this edge case.

Reply:
Passed the existing client in https://github.com/elastic/beats/pull/32785/commits/81b5df00de428016dd60cd79c51f1051455e5f58